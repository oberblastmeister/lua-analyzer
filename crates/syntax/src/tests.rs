use super::*;
use std::{fmt, fs, str};

fn dump_parse<T: AstNode + fmt::Debug>(parse: Parse<T>) -> String {
    let mut s = String::new();
    s.push_str(&format!("{:#?}", parse.syntax_node()));
    s.push_str(&format_errors(parse.errors()));
    s
}

fn dump_parse_no_errors(parse: Parse<Program>) -> Result<String, String> {
    if !parse.errors().is_empty() {
        return Err(dump_parse(parse));
        // panic!(
        //     "Should not have any errors {:#?}\n{}",
        //     parse.syntax_node(),
        //     format_errors(parse.errors())
        // )
    }
    Ok(parse.debug_dump())
}

#[test]
fn successes() {
    insta::glob!("snapshot_inputs/successes/*.lua", |path| {
        let input = fs::read_to_string(path).unwrap();
        let suffix = path.file_stem().unwrap().to_str().unwrap();
        insta::with_settings!({snapshot_path => "snapshots/successes", snapshot_suffix => suffix}, {
            let parse = Program::parse(&input);
            let res = dump_parse_no_errors(parse);
            match res {
                Ok(s) => insta::assert_snapshot!(s),
                Err(s) => panic!("Test {} should not have any errors:\n{}", path.display(), s),
            }
        })
    })
}

#[test]
fn luajit() {
    insta::glob!("snapshot_inputs/luajit/*.lua", |path| {
        let input = fs::read_to_string(path).unwrap();
        let suffix = path.file_stem().unwrap().to_str().unwrap();
        insta::with_settings!({snapshot_path => "snapshots/luajit", snapshot_suffix => suffix}, {
            let parse = Program::parse(&input);
            let res = dump_parse_no_errors(parse);
            match res {
                Ok(s) => insta::assert_snapshot!(s),
                Err(s) => panic!("Test {} should not have any errors:\n{}", path.display(), s),
            }
        })
    })
}

#[test]
fn fails() {
    insta::glob!("snapshot_inputs/fails/*.lua", |path| {
        let input = fs::read_to_string(path).unwrap();
        let suffix = path.file_stem().unwrap().to_str().unwrap();
        insta::with_settings!({snapshot_path => "snapshots/fails", snapshot_suffix => suffix}, {
            let parse = Program::parse(&input);
            insta::assert_snapshot!(dump_parse(parse));
        })
    })
}

#[test]
fn nothing() {
    insta::assert_debug_snapshot!(Program::parse("").syntax_node(), @"Program@0..0")
}

#[test]
fn fuzz1() {
    Program::parse("e,,");
}
