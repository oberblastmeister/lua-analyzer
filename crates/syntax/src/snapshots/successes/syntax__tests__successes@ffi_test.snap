---
source: crates/syntax/src/tests.rs
expression: s

---
Program@0..298
  Comment@0..14 "-- ffi.cdef[[\n"
  Comment@14..69 "-- typedef enum enum_ ..."
  Comment@69..123 "-- typedef enum enum_ ..."
  Whitespace@123..124 "\n"
  Comment@124..166 "-- enum_i call_ei_i(i ..."
  Comment@166..208 "-- enum_u call_eu_i(i ..."
  Comment@208..250 "-- int call_i_ei(enum ..."
  Comment@250..292 "-- int call_i_eu(enum ..."
  Comment@292..298 "-- ]]\n"

=============================
Errors:
=============================
[]
